### ISSUE TITLE
Multi-Label Datasets

### ISSUE THREAD

**Issue by harpalsahota** [KevinMusgrave/pytorch-metric-learning#231]
Hi There,

Firstly, great package! I have a question regarding multi-label datasets. I have a set of images that could belong to multiple classes and I cannot find in the documentation a way of training a model in a multi-label scenario? or a way of generating custom triplets which can then be passed to the loss function.

Thanks,

Harpal

**Comment by KevinMusgrave** [comment]
If your dataset returns labels in the form of a 2d numpy array, then the testers support multi-label evaluation. The base trainer can also be extended to support multi-label training. (See [this](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/178#issuecomment-675611566) for an example).

Custom triplets can be passed to the loss function like this:
```python
# assuming you've generated the triplet indices
indices_tuple = (anchors, positives, negatives)
loss = loss_fn(embeddings, labels, indices_tuple)
```

**Comment by harpalsahota** [comment]
Hey @KevinMusgrave,

Thanks for the swift reply. I'll take a look at what you said and try things on my end. I'm assuming the 2d numpy array you mentioned above would be the one-hot encoding of the labels?

**Comment by harpalsahota** [comment]
@KevinMusgrave does something like this make sense?

Define the loss function and miner
```pthon
miner = TripletMarginMiner(margin=0.2, type_of_triplets='hard')
loss_function = TripletMarginLoss(margin=0.1)
```

Training loop:
```python
        for i_batch, (images, product_labels) in enumerate(dataloader):
            optimizer.zero_grad()
            images = images.to(device)

            with torch.set_grad_enabled(phase == 'train'):
                image_emb = model(images)
                
            loss = 0.0
            # Iterate through each of the labels
            for i in range(product_labels.size(1)):
                current_labels = product_labels[:, i]
                hard_pairs = miner(image_emb, current_labels)
                loss += loss_function(image_emb, current_labels, hard_pairs)
            loss /= product_labels.size(1)
```

**Comment by KevinMusgrave** [comment]
Yeah I think that makes sense

**Comment by priyarana** [comment]
> ```python
> loss = 0.0
> ```

@harpalsahota  Hi. . My data is multi label as well and i am struggling with metric learning.  Just wondering, did this work for you ?. Thanks

**Comment by Mahiro2211** [comment]
Thanks a lot 🙏
